Fix for issue 2502: InfluxDB binding throws (uncaught) NumberFormatException when value is string

Chel opened the pull request:
This allows the retrieval of String items from the InfluxDB. Tested against XBMC String items.

buildhive commented:
[openhab » openhab #2861](https://buildhive.cloudbees.com/job/openhab/job/openhab/2861/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

theoweiss commented:
Hi Chel,

after a closer look into the persistence services I think your fix is worth fixing the bug you observed! 
But the influxdb service is missing the handling for additional Item types. I looked at the mysql and mongodb services they handle the following mappings:

Mysql:
NumberItem -> DecimalType
ColorItem -> HSBType
SwitchItem -> OnOffType
ContactItem -> RollershutterItem
DimmerItem -> PercentType
RollershutterItem -> PercentType
DateTimeItem -> DateTimeType
else -> StringType

MongoDB:
NumberItem -> DecimalType
DimmerItem -> PercentType
SwitchItem -> OnOffType
ContactItem -> OpenClosedType
RollershutterItem -> PercentType
ColorItem -> HSBType
DateTimeItem -> DateTimeType
else -> StringType

The influxdb service should be enhanced accordingly. Furthermore the new LocationItem should be handled: LocationItem -> PointType. (@teichsta I think some other persistance services should also be changed to handle the LocationItem.)

The over all fallback should be StringType because no further conversion is needed.
Please keep the type hierachy in mind: ColorItem -> DimmerItem -> SwitchItem.

Would you mind to enhance the influxdb service accordingly?

May be after 1.7 a more fundamentally change would be good: may be we could use item.getAcceptedDataTypes() and inspect the value retrieved from the db to guess the best data type to be returned. We should have some thoughts from @teichsta or @kaikreuzer about this.


Thanks,
Theo

Chel commented:
Theo, thanks for the in-depth review comments. I will make the enhancements you suggest, tomorrow evening if I am able.

teichsta commented:
> @teichsta I think some other persistance services should also be changed to handle the LocationItem.

good catch @theoweiss will announce that on the list and hope to find volunteers for that job :-)

> may be we could use item.getAcceptedDataTypes() and inspect the value retrieved from the db to guess the best data type to be returned.

sounds like a general issue which should be discussed in ESH?

teichsta commented:
any chance to get this merged until the weekend (May, 16th) @theoweiss?

theoweiss commented:
@Chel what is the status of your efforts? Do you need some help?

Chel commented:
Hi Theo, real life got the better of me this week I'm afraid. Give me 24 hours and I should have something suitable.

theoweiss commented:
Hi @Chel, sounds great!

teichsta commented:
Thanks guys! Good to know you'll take care of this …

theoweiss commented:
Hi @Chel, is there progress? We should have a solution this evening.

Chel commented:
Hi, I am actively looking at this this evening but it's taking a while - the class has an implicit assumption that all values will be numeric, and unteasing that is taking a while due to my lack of familiarity (though that is increasing!). I may get a check-in done before bed but it is more likely to be tomorrow.

Chel commented:
Okay, I thought my commit last night was unsuccessful; (I'm having a lot of trouble figuring out how git/github works, especially via the eclipse plugins), but it seems that my commit did work, and is above. Can somebody review please? I am available this weekend for further rework as necessary.

theoweiss commented:
Thanks @Chel. I will do the review this evening/night.

buildhive commented:
[openhab » openhab #2970](https://buildhive.cloudbees.com/job/openhab/job/openhab/2970/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

buildhive commented:
[openhab » openhab #2972](https://buildhive.cloudbees.com/job/openhab/job/openhab/2972/) FAILURE
Looks like there's a problem with this pull request
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

theoweiss commented:
I @Chel, I have almost completed the review, I will post my comments tomorrow.

theoweiss commented on bundles/persistence/org.openhab.persistence.influxdb/java/org/openhab/persistence/influxdb/internal/InfluxDBPersistenceService.java:
```diff
@@ -415,24 +421,31 @@ private Object convertBigDecimalToNum(BigDecimal value) {
-      value = ((DateTimeType) state).getCalendar().getTime().getTime();
+    if (state instanceof DecimalType) {
+    	value = convertBigDecimalToNum(((DecimalType) state).toBigDecimal());
+    } else if (state instanceof PercentType) {
```
Remove handling of PercentType because it is a subtype of DecimalType and both are converted in the same way.

theoweiss commented on bundles/persistence/org.openhab.persistence.influxdb/java/org/openhab/persistence/influxdb/internal/InfluxDBPersistenceService.java:
```diff
@@ -415,24 +421,31 @@ private Object convertBigDecimalToNum(BigDecimal value) {
-      value = (OpenClosedType) state == OpenClosedType.OPEN ? 1 : 0;
+    	value = (OpenClosedType) state == OpenClosedType.OPEN ? 1 : 0;
+    } else if (state instanceof HSBType) {
+    	value = ((HSBType) state).getConstituents();
```
Use ((HSBType) state).toString(), we should not rely on the toString method of the AbstractMap.

theoweiss commented on bundles/persistence/org.openhab.persistence.influxdb/java/org/openhab/persistence/influxdb/internal/InfluxDBPersistenceService.java:
```diff
@@ -415,24 +421,31 @@ private Object convertBigDecimalToNum(BigDecimal value) {
+    } else if (state instanceof HSBType) {
+    	value = ((HSBType) state).getConstituents();
+    } else if (state instanceof DateTimeType) {
+    	value = ((DateTimeType) state).getCalendar().getTime().getTime();
```
((DateTimeType) state).toString() should do the trick

theoweiss commented on bundles/persistence/org.openhab.persistence.influxdb/java/org/openhab/persistence/influxdb/internal/InfluxDBPersistenceService.java:
```diff
@@ -415,24 +421,31 @@ private Object convertBigDecimalToNum(BigDecimal value) {
+    } else if (state instanceof DateTimeType) {
+    	value = ((DateTimeType) state).getCalendar().getTime().getTime();
+    } else if (state instanceof PointType) {
+    	value = ((PointType) state).getConstituents();
```
We need a converter for PointType to String. Unfortunately ((PointType) state).toString() is odd because we can not feed this value back into a PointType constructor. I think a concatenated comma separated String of latitude, longitude, altitude should work. On restore this string has to be converted to a valid input for a PointType constructor. A similar approach can be found in HSBType constructor HSBType(String value).

theoweiss commented on bundles/persistence/org.openhab.persistence.influxdb/java/org/openhab/persistence/influxdb/internal/InfluxDBPersistenceService.java:
```diff
@@ -445,16 +458,22 @@ private Object stateToObject(State state) {
+    if (state instanceof DecimalType) {
+    	value = ((DecimalType) state).toBigDecimal().toString();
+    } else if (state instanceof PercentType) {
+    	value = ((PercentType) state).toBigDecimal().toString();
```
PercentType is a DecimalType and for DecimalType calling toString() is the right thing.

theoweiss commented on bundles/persistence/org.openhab.persistence.influxdb/java/org/openhab/persistence/influxdb/internal/InfluxDBPersistenceService.java:
```diff
@@ -445,16 +458,22 @@ private Object stateToObject(State state) {
-      value = ((OnOffType) state) == OnOffType.ON ? DIGITAL_VALUE_ON : DIGITAL_VALUE_OFF;
+    	value = ((OnOffType) state) == OnOffType.ON ? DIGITAL_VALUE_ON : DIGITAL_VALUE_OFF;
+    } else if (state instanceof OpenClosedType) {
+    	value = ((OpenClosedType) state) == OpenClosedType.OPEN ? DIGITAL_VALUE_OPEN : DIGITAL_VALUE_CLOSED;  
```
because we used 0/1 for storing the OpenClosedType (line 440) we must also use it for querying. Use DIGITAL_VALUE_ON/OFF here.

theoweiss commented on bundles/persistence/org.openhab.persistence.influxdb/java/org/openhab/persistence/influxdb/internal/InfluxDBPersistenceService.java:
```diff
@@ -445,16 +458,22 @@ private Object stateToObject(State state) {
+    } else if (state instanceof OpenClosedType) {
+    	value = ((OpenClosedType) state) == OpenClosedType.OPEN ? DIGITAL_VALUE_OPEN : DIGITAL_VALUE_CLOSED;  
+    } else if (state instanceof HSBType) {
+    	value = ((HSBType) state).getConstituents().toString();
```
use ((HSBType) state).toString()

theoweiss commented on bundles/persistence/org.openhab.persistence.influxdb/java/org/openhab/persistence/influxdb/internal/InfluxDBPersistenceService.java:
```diff
@@ -445,16 +458,22 @@ private Object stateToObject(State state) {
+    } else if (state instanceof HSBType) {
+    	value = ((HSBType) state).getConstituents().toString();
+    } else if (state instanceof DateTimeType) {
+    	value = String.valueOf(((DateTimeType) state).getCalendar().getTime().getTime());
```
use ((DateTimeType) state).toString()

theoweiss commented on bundles/persistence/org.openhab.persistence.influxdb/java/org/openhab/persistence/influxdb/internal/InfluxDBPersistenceService.java:
```diff
@@ -445,16 +458,22 @@ private Object stateToObject(State state) {
+    } else if (state instanceof DateTimeType) {
+    	value = String.valueOf(((DateTimeType) state).getCalendar().getTime().getTime());
+    } else if (state instanceof PointType) {
+    	value = ((PointType) state).getConstituents().toString();
```
Use a PointType to String converter as mentioned in my previous comment.

theoweiss commented on bundles/persistence/org.openhab.persistence.influxdb/java/org/openhab/persistence/influxdb/internal/InfluxDBPersistenceService.java:
```diff
@@ -465,46 +484,87 @@ private String stateToString(State state) {
+        if (item instanceof ColorItem || item instanceof DimmerItem) {
+        	return item.getState();
+        } else if (item instanceof SwitchItem) {
+        	return convertToSwitchOutput(valueStr).equals(DIGITAL_VALUE_OFF)
```
Replacing string2DigitalValue with convertToSwitchOutput and convertToContactOutput is a bad idea because OpenClosedType is stored as 0 or 1. Furthermore I prefer the overhead of the Bigdecimal approach over the exception handling of the Integer.parseInt().

theoweiss commented on bundles/persistence/org.openhab.persistence.influxdb/java/org/openhab/persistence/influxdb/internal/InfluxDBPersistenceService.java:
```diff
@@ -415,24 +421,31 @@ private Object convertBigDecimalToNum(BigDecimal value) {
+   *         	integer for DateTimeType,
+   *         	SortedMap<String, PrimitiveType> for PointType
    */
   private Object stateToObject(State state) {
```
stateToObject is called from line 192. You should use getStateAs for DimmerItem/RollershutterItem and ColorItem
have a look at https://github.com/openhab/openhab/blob/master/bundles/persistence/org.openhab.persistence.mysql/java/org/openhab/persistence/mysql/internal/MysqlPersistenceService.java#L306

theoweiss commented:
Code formatting: please use spaces for indentation, current indentation size is 2.
@teichsta may be we should use the ESH formatter also for openHAB?

theoweiss commented:
@Chel Please find my review comments inline. Let me know when your ready or something has to be discussed.
Thanks.

buildhive commented:
[openhab » openhab #2987](https://buildhive.cloudbees.com/job/openhab/job/openhab/2987/) FAILURE
Looks like there's a problem with this pull request
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

Chel commented:
I've addressed all of your comments. Sorry for the noise, but I replaced all 'tabs' with spaces to ensure mine was right. Regarding the change to PointType.java, have I understood correctly?

I'm not sure if the buildhive failure is relevant - considering the time it took to return, I'm suspicious that the error is not with my code, but with the build process.

Chel commented:
@theoweiss Just in case you're not subscribed.

theoweiss commented:
Thanks Chel. I can do the review this evening, hopefully not to late.
Be prepared to add yourself to the authors list and squashing your commits into one. But wait with the squashing until I've done the review.

Chel commented:
That's cool, thanks. I don't know how to squash but I'm reading up on it now.

theoweiss commented:
Sorry for being unclear regarding the PointType. The converter should be a part of the Influxdb service. The cause of the buildhive failure is a conflicting change to the PointType's toString method. Yesterday Thomas merged a change. Could you please revert your changes to PointType and write the converter.

Chel commented:
@theoweiss  Done, please take a look.

For the record I don't agree that the string conversion should be done in the InfluxDB service, I think it makes much more sense design-wise to offer it in the PointType, perhaps as toUnlabelledString(). Doing the conversion in the InfluxDB class is messy and prone to error, in my opinion. That said, I've done it as you asked.

Further, I'm wondering if there is an error in the PointType.toString() method, as it always returns '°N' and '°W', never '°S' or '°E'. I understand that the northings/eastings can be negative, but I ask as the PointType.canonicalize() class matches that found at the following link, and you'll see that the toString() method in the link does return S and E, when appropriate: https://collab.firelab.org/svn/modisfire/jts-kd-tree/tags/pre-maven/geodesy/src/org/gavaghan/geodesy/GlobalCoordinates.java

So, my converter doesn't handle S or E, as they can't be returned. Does what I've done look like what you wanted, or did you have some other manner in mind? I'm reaching the end of my available time today, so any substantial rework may have to wait until Wednesday evening (I've a job interview on Tuesday to prepare for).

buildhive commented:
[openhab » openhab #2990](https://buildhive.cloudbees.com/job/openhab/job/openhab/2990/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

theoweiss commented:
Sorry again, if my comments about the PointType were misleading! The PR looks good, despite the PointType handling. I think the PointType handling should be solved in a subsequent PR so that we can get this one ready for 1.7. Could you please remove the PointType handling, I will open up a new PR for that issue.
All the best for your job interview!

Chel commented:
Thanks  :O)   I'll be able to get to this on Wednesday evening, but I'm not 100% clear on what you are asking. Do you want me to delete the removeLabelsFromPointType() method, and rely on PointType.toString() returning an unlabelled string? 

Or are you asking me to remove all mention of PointType?

theoweiss commented:
Please remove all mentions of PointType, we will handle this with a  new PR.

FYI: If your unsure how squashing works Kai recently posted a hint: https://github.com/ginatrapani/todo.txt-android/wiki/Squash-All-Commits-Related-to-a-Single-Issue-into-a-Single-Commit

Chel commented:
Ah thanks. I had tried the 'git rebase -i HEAD~4' command the other day, but I get the following error:

    dan@attic:/opt/openhab-dev/openhab$ git rebase -i HEAD~4
    Cannot rebase: You have unstaged changes.
    Please commit or stash them.
    dan@attic:/opt/openhab-dev/openhab$

I think the unstaged changes are the build of the model that we need to do when setting up the IDE, is that right? If so I'd expect everybody to get this error, so I'm not sure. Anyway, I'll look at it throught he week. Thanks again for all of your patience.

teichsta commented:
> I think the unstaged changes are the build of the model that we need to do

use ```git status``` to see the files in question …

Chel commented:
@teichsta thanks, I needed to 'stash' the local model changes.
@theoweiss I've done what was asked and squashed, however I've brought 3 extra commits with me. As they don't carry file changes, is this okay, or do I need to do something to remove them?

buildhive commented:
[openhab » openhab #3001](https://buildhive.cloudbees.com/job/openhab/job/openhab/3001/) FAILURE
Looks like there's a problem with this pull request
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

theoweiss commented:
Chel you have screed up your branch, there are 69 changed files. Is there a chance to get this fixed? Otherwise close this PR and open a new one with a clean state.

Chel commented:
You're right, I'm terrible at github  :O(

I've deleted my repo (with my changes backed up separately) and I'm going to attempt to start from scratch with just the relevant changes.

theoweiss commented:
Ok. I close this pull request. Open a new one from your new repo. To give you a tip: don't work on the master branch, use separate branches for your dev. If something goes wrong you only have to delete the branch not the whole repo.
